### PR TITLE
[feature] Adyen 3DS support

### PR DESCRIPTION
**Pull request by dbackowski**
This adds support for Adyen 3DS

```
Loaded suite test/remote/gateways/remote_adyen_checkout_test
Started
..................
Finished in 7.349448 seconds.
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
18 tests, 50 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

```
Loaded suite test/unit/gateways/adyen_checkout_test
Started
...............
Finished in 0.063384 seconds.
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
15 tests, 82 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

Co-authored-by: nbialostosky <nicholas.bialostosky@chargify.com>
Co-authored-by: Przemysław Haptaś <przemyslaw.haptas@chargify.com>

### CODE REVIEW

File: lib/active_merchant/billing/gateways/adyen_checkout.rb [hunk 1]
```diff
@@ -153,6 +162,19 @@ def add_risk_data(post, options)
         end
       end
 
+      def add_3ds_data(post, options)
+        post[:additionalData] ||= {}
+        post[:additionalData][:allow3DS2] = options[:allow3DS2] if options[:allow3DS2]
+        post[:channel] = options[:channel] if options[:channel]
+        post[:origin] = options[:origin] if options[:origin]
+        post[:browserInfo] = options[:browser_info] if options[:browser_info] # Autopopulated with `populateBrowserInfoFor3ds` service enabled
```
Comment (przemyslawhaptas) [inline lib/active_merchant/billing/gateways/adyen_checkout.rb:170]: Have you talked with Adyen support about this service? Perhaps we don't need to do all of that https://github.com/chargify/chargify/pull/15768/files#diff-676f6becf571acae78e6f3195c34bed8R9?

If we do, though, let's remove this comment.

Comment (przemyslawhaptas) [inline lib/active_merchant/billing/gateways/adyen_checkout.rb:170]: Have you talked with Adyen support about this service? Perhaps we don't need to do all of that https://github.com/chargify/chargify/pull/15768/files#diff-676f6becf571acae78e6f3195c34bed8R9?

If we do, though, let's remove this comment.

File: lib/active_merchant/billing/gateways/adyen_checkout.rb [hunk 2]
```diff
@@ -69,6 +72,12 @@ def store(credit_card, options={})
         end
       end
 
+      def details(options)
```
Comment (przemyslawhaptas) [inline lib/active_merchant/billing/gateways/adyen_checkout.rb:75]: We probably should have some spec for that.

File: lib/active_merchant/billing/gateways/adyen_checkout.rb [hunk 3]
```diff
@@ -52,13 +52,16 @@ def refund(money, authorization, options={})
       end
 
       def store(credit_card, options={})
+        return details(options) if options[:three_ds_data]
```
Comment (przemyslawhaptas) [inline lib/active_merchant/billing/gateways/adyen_checkout.rb:55]: it seems we're using three different names for the same thing in this file: `three_ds`, `3ds`, `3DS2`. Can we do something about it?

File: lib/active_merchant/billing/gateways/adyen_checkout.rb (hunk 1 above)
Comment (dbackowski) [reply lib/active_merchant/billing/gateways/adyen_checkout.rb:170]: I didn't talk to them, but if I understand correctly this browser's data are passed to onChange event that we don't use, this event is never fired. I found that they are providing some utils https://github.com/Adyen/adyen-3ds2-js-utils, so I just grabbed their collectBrowserInfo implementation

File: lib/active_merchant/billing/gateways/adyen_checkout.rb (hunk 2 above)
Comment (dbackowski) [reply lib/active_merchant/billing/gateways/adyen_checkout.rb:75]: We have tests in Chargify that test the whole flow with 3DS. Writing tests here for 3d secure flow is very hard, so probably we would have to stub everything and I don't think that this is the best approach.

File: lib/active_merchant/billing/gateways/adyen_checkout.rb (hunk 3 above)
Comment (dbackowski) [reply lib/active_merchant/billing/gateways/adyen_checkout.rb:55]: I see only `options[:three_ds_data]` and methods like `add_3ds_data` and `add_3ds_details`, there is also `allow3DS2` but this is an option from Adyen that we need to set to true in some cases (so we can't change the name of it).
I've changed this `add_3ds_data` into `add_three_ds_data` and `add_3ds_details` into `add_three_ds_details`

Comment (przemyslawhaptas) [reply lib/active_merchant/billing/gateways/adyen_checkout.rb:55]: one fewer, cool! 🎉

File: lib/active_merchant/billing/gateways/adyen_checkout.rb (hunk 2 above)
Comment (przemyslawhaptas) [reply lib/active_merchant/billing/gateways/adyen_checkout.rb:75]: @himn1 you'll probably be the one approving the release so you could as well tell us now what our policy is. How do we even treat AM? Is it our own gem at this point? Or are w planning on somehow joining the main fork at some point in the future?

Comment (dbackowski) [reply lib/active_merchant/billing/gateways/adyen_checkout.rb:75]: I've added some unit tests

File: test/fixtures.yml [hunk 4]
```diff
@@ -17,9 +17,9 @@ adyen:
   merchant_account: ''
 
 adyen_checkout:
-  username: 'ws@Company.Chargify692'
-  password: '@t2)a2wp74cP^5-.N+VsDIM8g'
-  merchant_account: 'Chargify692ECOM'
```
Comment (przemyslawhaptas) [inline test/fixtures.yml]: 😅

File: lib/active_merchant/billing/gateways/adyen_checkout.rb [hunk 5]
```diff
@@ -220,8 +220,6 @@ def add_shopper_interaction(post, payment, options={})
       end
 
       def add_recurring_processing_model(post, options)
-        return unless options.dig(:stored_credential, :reason_type) || options[:recurring_processing_model]
```
Comment (przemyslawhaptas) [inline lib/active_merchant/billing/gateways/adyen_checkout.rb:201]: How do we know that for any purchase ever we should set the `post[:recurringProcessingModel]` flag?
I know removing this guard solves the PSP case but is it the right solution?

File: test/fixtures.yml (hunk 4 above)
Comment (dbackowski) [reply test/fixtures.yml]: 🤦‍♂

File: lib/active_merchant/billing/gateways/adyen_checkout.rb (hunk 5 above)
Comment (dbackowski) [reply lib/active_merchant/billing/gateways/adyen_checkout.rb]: OK, I will restore this, and I will pass `recurringProcessingModel` as "Subscription" from the conduit

File: lib/active_merchant/billing/gateways/adyen_checkout.rb (hunk 2 above)
Comment (himn1) [reply lib/active_merchant/billing/gateways/adyen_checkout.rb:75]: Answered in-slack
recap:
Definitely we have to make an effort soon and try to be upstream.
Step after step with a proper plan.

**Pull request merged**